### PR TITLE
feat: add more condition and make cpl concurrent install

### LINKED ISSUE #85: parallel installation of componentplan(helm lock)

### LINKED ISSUE #117: unit tests failed in helm package
See  https://github.com/kubebb/core/actions/runs/5572505081/jobs/10178619907
```text
--- FAIL: Test_postRenderer_Run (0.25s)
    --- FAIL: Test_postRenderer_Run/override_registry,_kustomize_image_and_define_componentplan_name (0.05s)
        post_renderer_test.go:213: Run() gotModifiedManifests = apiVersion: v1
            kind: Pod
            metadata:
              labels:
                app: nginx
              name: nginx
              namespace: default
            spec:
              containers:
              - image: docker.cc/library/nginx:v2
                imagePullPolicy: IfNotPresent
                name: nginx
            ---
            apiVersion: apps/v1
            kind: Deployment
            metadata:
              labels:
                control-plane: controller-manager
              name: controller-manager
              namespace: baas-system
            spec:
              replicas: 1
              selector:
                matchLabels:
                  control-plane: controller-manager
                  name: controller-manager
              template:
                metadata:
                  labels:
                    control-plane: controller-manager
                    name: controller-manager
                spec:
                  containers:
                  - image: 172.22.50.223/bestchains-dev/fabric-operator:7776e71
                    name: operator
            , want apiVersion: v1
            kind: Pod
            metadata:
              labels:
                app: nginx
                core.kubebb.k8s.com.cn/componentplan: abcd
              name: nginx
              namespace: default
            spec:
              containers:
              - image: docker.cc/library/nginx:v2
                imagePullPolicy: IfNotPresent
                name: nginx
            ---
            apiVersion: apps/v1
            kind: Deployment
            metadata:
              labels:
                control-plane: controller-manager
                core.kubebb.k8s.com.cn/componentplan: abcd
              name: controller-manager
              namespace: baas-system
            spec:
              replicas: 1
              selector:
                matchLabels:
                  control-plane: controller-manager
                  core.kubebb.k8s.com.cn/componentplan: abcd
                  name: controller-manager
              template:
                metadata:
                  labels:
                    control-plane: controller-manager
                    core.kubebb.k8s.com.cn/componentplan: abcd
                    name: controller-manager
                spec:
                  containers:
                  - image: 172.22.50.223/bestchains-dev/fabric-operator:7776e71
                    name: operator
        post_renderer_test.go:215: diff: metadata:
              labels:
                core.kubebb.k8s.com.cn/componentplan: abcd -> <empty> (REMOVED)
FAIL
	github.com/kubebb/core/pkg/helm	coverage: 9.1% of statements
FAIL	github.com/kubebb/core/pkg/helm	0.[43](https://github.com/kubebb/core/actions/runs/5572505081/jobs/10178619907?pr=110#step:4:44)2s
```

### PR DESCRIPTION
**Pull request by Abirdcfly**
Fix #85 
the second commit xref #105
Fix #117

### CODE REVIEW

**Comment by Abirdcfly** [comment]
@CodiumAI-Agent

**Comment by Abirdcfly** [comment]
@CodiumAI-Agent

File: api/v1alpha1/componentplan.go [hunk 1]
```diff
@@ -110,6 +132,18 @@ func componentPlanCondition(ct ConditionType, reason ConditionReason, status cor
 	return c
 }
 
+func (c *ComponentPlan) InitCondition() []Condition {
+	if c.Spec.Approved {
+		return []Condition{ComponentPlanAppreoved(), ComponentPlanWaitDo(nil), ComponentPlanInitSucceded()}
```
Comment (bjwswang) [inline api/v1alpha1/componentplan.go:137]: type here `ComponentPlanAppreoved`

File: api/v1alpha1/componentplan.go [hunk 2]
```diff
@@ -30,15 +30,21 @@ const (
 const (
 	ComponentPlanTypeSucceeded ConditionType = "Succeeded"
 	ComponentPlanTypeApproved  ConditionType = "Approved"
-	ComponentPlanTypeInstalled ConditionType = "Installed"
+	ComponentPlanTypeActioned  ConditionType = "Actioned"
```
Comment (bjwswang) [inline api/v1alpha1/componentplan.go:33]: explain this `Actioned` condtion type

File: controllers/componentplan_controller.go [hunk 3]
```diff
@@ -143,29 +143,46 @@ func (r *ComponentPlanReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+			logger.Info("the last one is uninstalling... skip for 1 minute")
+			return ctrl.Result{RequeueAfter: time.Minute}, nil
 		}
-
-		logger.Info("Removing Finalizer for ComponentPlan after successfully performing the operations")
-		if ok := controllerutil.RemoveFinalizer(plan, corev1alpha1.Finalizer); !ok {
-			return ctrl.Result{Requeue: true, RequeueAfter: 3 * time.Second}, nil
+		if plan.InActionedReason(corev1alpha1.ComponentPlanReasonUnInstallFailed) {
+			// TODO do we need uninstall retry like install or upgrade done?
+			logger.Info("the last one is uninstallfailed, just skip this one")
```
Comment (bjwswang) [inline controllers/componentplan_controller.go:152]: good point! we can do `uninstall retry` here but with maybe only `one more time???`

File: api/v1alpha1/componentplan.go (hunk 2 above)
Comment (Abirdcfly) [reply api/v1alpha1/componentplan.go:33]: A **clearer** name for what has been operated may be needed here...
The specifics of the operation will be described by `ConditionReason`, including:
- WaitDo
- Install: Installing / InstallSuccess / InstallFailed
- Uninstall: UnInstalling / UnInstallSuccess / UnInstallFailed
- Upgrade: Upgrading / UpgradeSuccess / UpgradeFailed

one example will be:
```yaml
  status:
    conditions:
    - lastTransitionTime: "2023-07-16T07:25:44Z"
      reason: ""
      status: "True"
      type: Approved
    - lastTransitionTime: "2023-07-16T07:26:14Z"
      reason: Upgrading
      status: "False"
      type: Actioned
    - lastTransitionTime: "2023-07-16T07:25:44Z"
      reason: ""
      status: "False"
      type: Succeeded
```

**Review by bjwswang — APPROVED** [review]
pls fix the typos and add some comments

File: api/v1alpha1/componentplan.go (hunk 2 above)
Comment (bjwswang) [reply api/v1alpha1/componentplan.go:33]: 👌  we should explain `CondtionType` in our document later.

Comment (Abirdcfly) [reply api/v1alpha1/componentplan.go:33]: will track in #119

File: controllers/componentplan_controller.go (hunk 3 above)
Comment (Abirdcfly) [reply controllers/componentplan_controller.go:152]: will track in #120

**Pull request merged**